scripts/imgtool.py

```diff
@@ -1,15 +1,33 @@
 #! /usr/bin/env python3
 
 import argparse
+import getpass
 from imgtool import keys
 from imgtool import image
 from imgtool import version
 import sys
 
+def get_password(args):
+    if args.password:
+        while True:
+            passwd = getpass.getpass("Enter key passphrase: ")
+            passwd2 = getpass.getpass("Reenter passphrase: ")
+            if passwd == passwd2:
+                break
+            print("Passwords do not match, try again")
+
+        # Password must be bytes, assume UTF-8.
+        return passwd.encode('utf-8')
```

Comment:
Why not sys.stdin.encoding? Where else does getpass get its input from?

(Same comment applies elsewhere)

Reply:
The "correct" answer is to file a bug against `getpass`. It should return a byte-string, and not be decoding/reencoding a string.

It is pretty reasonable to assume the user's terminal is UTF-8. There are too many other things that wouldn't be working for them if that weren't the case, including all of Go.

Reply:
> The "correct" answer is to file a bug against getpass. It should return a byte-string, and not be decoding/reencoding a string.

No, a password is inherently a human-defined string. That means Unicode, not bytes. Just because cryptography.hazmat.primitives.serialization.BestAvailableEncryption wants bytes doesn't change that.

So, I read the getpass.py source code for Python 3.6.

On Unix, getpass tries:

- TextIOWrapper(FileIO('/dev/tty')).readline(), which returns str
- sys.stdin.readline(), which returns str

On Windows, it uses msvcrt.getwch() (https://docs.python.org/3/library/msvcrt.html#msvcrt.getwch), which returns "a Unicode value". I'm assuming that starts with the wint_t returned from conio.h's _getwch() under the covers.

So getting a Unicode value back from getpass() makes sense.

> It is pretty reasonable to assume the user's terminal is UTF-8.

I'm not sure what the terminal has to do with anything, especially since two out of the three possible input sources for getpass() are not (necessarily) Unix terminals.

The real issue here seems to be that the cryptography module wants a byte string argument to BestAvailableEncryption. I assume you ran into the `raise ValueError("Password must be 1 or more bytes.")` in serialization.py. That has nothing to do with the user's terminal as far as I can tell.

> There are too many other things that wouldn't be working for them if that weren't the case, including all of Go.

That seems like a non-sequitur. Go is not a mandatory mcuboot dependency in the Zephyr port. What are you saying?

At any rate, I think you really should reconsider this in favor of something like sys.getdefaultencoding(). That way we both get what we seem to want: the value stays utf-8 usually, but the code isn't making assumptions. Remember the Zen of Python:

"In the face of ambiguity, refuse the temptation to guess."

Reply:
> At any rate, I think you really should reconsider this in favor of something like sys.getdefaultencoding().

And perhaps filing a bug against cryptography, not getpass :wink:

Reply:
Actually, after some thought, I will argue that this should _always_ be using UTF-8. The fact that the user may have different encodings at different times only argues more strongly that we should always use a consistent encoding when using this input as a password.

Let's say a user starts out with Shift-JIS encoding. As long as getpass does things right, the result will end up as a Unicode string within python, because in Python 3, strings are always Unicode. Later, they may change their configuration to use a UTF-8-based input system. The string should still come into python as Unicode.

If we used the user's encoding, the password bytes would be different. But, if we consistently use UTF-8, we'll get the same password bytes, no matter how the user input it.

Reply:
Alright, that's fair. The comment you've added is a good explanation for the updated rationale.